### PR TITLE
Core: Rollback compaction on conflicts

### PR DESCRIPTION
**Pull request by singhpk234**
### About the change : 
Revert compaction in case a conflict is detected with the non compaction writes.

* Identify commits  created via compaction, we can use snapshot's `operation()` to check if the given snapshot is of `REPLACE` if yes then it was created as part of compaction.

* While committing the pending updates of the transaction, check if the current snapshot conflicts with the transaction updates. If the current snapshot conflicts and it’s a snapshot created by the compaction process (snapshot#operation()). Then revert the current snapshot (rollback to its parent snapshot) and then try re-applying updates on top of it.
Introduce a new table property `rollback.compaction.on-conflicts.enabled`, which essentially helps in identifying if we want to rollback the compaction commit if conflicts are detected.

* Let’s say a transaction has U1, U2, U3 ... update to be applied, let say B is the base snapshot, now when we were going to actually commit the transaction it saw the current snapshot now is B`(a snapshot created due to compaction), so make current snapshot as B` and try re-apply updates i.e {U1, U2, U3}, when applying U3 on top of { B` -> U1 -> U2 } a metadata conflict was detected, then try updates of transactions by rolling back B` to it’s parent, Let's say rollback B` to its parent is called RollbackB` so now try applying {RollbackB`, U1, U2, U3} on top of B` and see if it still conflicts if yes fail, else commit and update table metadata.

An alternative to proposal : https://docs.google.com/document/d/1pSqxf5A59J062j9VFF5rcCpbW9vdTbBKTmjps80D-B0/edit# suggested in community sync.

Adding this here to get a feedback on the approach. Once agreed upon, can make engines (for ex: spark) use transaction api.

-----

### Testing done : 
- Added ut's for validating is_compacted is present in compacted snapshot summary
- Added ut for e2e using transaction API and concurrent rewrite.
- TODO add more exhaustive UT's 

cc @rdblue @danielcweeks @jackye1995 @rajarshisarkar @amogh-jahagirdar

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 1]
```diff
@@ -479,6 +480,52 @@ private void applyUpdates(TableOperations underlyingOps) {
...
+            // use refreshed metadata
+            this.base = underlyingOps.current();
+            this.current = underlyingOps.current();
+            Long rollbackToSnapshotId = current.currentSnapshot().parentId();
+            long currentSnapshotId = current.currentSnapshot().snapshotId();
+            boolean updatesAppliedSuccessfully = false;
+
+            while (rollbackToSnapshotId != null
+                && !updatesAppliedSuccessfully
+                && PropertyUtil.propertyAsBoolean(
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:499]: I discussed with you a bit offline as well, I think instead of introducing a new property of `is-compacted`, it seems like we can rollback whenever we have conflict and `current.snapshot(currentSnapshotId).equals(DataOperations.REPLACE)`? Which applies to both compaction and rewrite manifests and feels more generic. Is there any risk of doing that way?

Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java]: >  I discussed with you a bit offline as well, I think instead of introducing a new property of is-compacted, it seems like we can rollback whenever we have conflict and current.snapshot(currentSnapshotId).equals(DataOperations.REPLACE)

Apologies, I missed that. I think it makes sense to use snapshot#operation to identify if it's a compaction. Updated the PR & description.

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 2]
```diff
@@ -479,6 +480,49 @@ private void applyUpdates(TableOperations underlyingOps) {
           // Cannot pass even with retry due to conflicting metadata changes. So, break the
           // retry-loop.
           throw new PendingUpdateFailedException(e);
+
+        } catch (ValidationException e) {
+          if (PropertyUtil.propertyAsBoolean(
+              base.properties(),
+              TableProperties.ROLLBACK_COMPACTION_ON_CONFLICTS_ENABLED,
+              TableProperties.ROLLBACK_COMPACTION_ON_CONFLICTS_ENABLED_DEFAULT)) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:488]: thanks for the update! In that case should we call this something like `ROLLBACK_REWRITE_ON_CONFLICTS_ENABLED`?

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java:488]: also the term rollback seems to be not accurate. We are not just rolling back, but we are also reapplying the changes after rolling back. Maybe something like `REAPPLY_REPLACE_SNAPSHOT_ON_CONFLICTS_ENABLED`?

Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java:488]: > We are not just rolling back, but we are also reapplying the changes after rolling back. Maybe something like REAPPLY_REPLACE_ON_CONFLICTS_ENABLED? Any other suggestions for the name?

makes sense, how about something like `RETRY_WITH_ROLLBACK_REPLACE_ON_CONFLICTS_ENABLED`, it gives an impression that we are retrying the applying the changes with roll-backing replace operations snapshot.

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 3]
```diff
@@ -466,6 +467,7 @@ private void cleanUpOnCommitFailure() {
         .run(ops.io()::deleteFile);
   }
 
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:470]: Rather than supressing this, it's a good sign that the new code should be refactored.

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 4]
```diff
@@ -479,6 +481,55 @@ private void applyUpdates(TableOperations underlyingOps) {
...
+            this.current = underlyingOps.current();
+            Long rollbackToSnapshotId = current.currentSnapshot().parentId();
+            long currentSnapshotId = current.currentSnapshot().snapshotId();
+            boolean updatesAppliedSuccessfully = false;
+
+            while (rollbackToSnapshotId != null
+                && !updatesAppliedSuccessfully
+                && DataOperations.REPLACE.equals(current.snapshot(currentSnapshotId).operation())) {
+              SetSnapshotOperation setSnapshotOp = new SetSnapshotOperation(transactionOps);
+              setSnapshotOp.rollbackTo(rollbackToSnapshotId);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:502]: Doesn't this need to be committed?

**Review by rdblue — CHANGES_REQUESTED** [review]
I think that we need to change the approach in this change. The current PR updates transactions, but that is a harder problem because there are multiple commits to handle. It also performs a somewhat blind rewrite by just checking to see if rolling back works, rather than going back to a starting snapshot. That seems really risky to me.

I think that this work should be done in `SnapshotProducer` rather than in `BaseTransaction`. That will work for direct commits as well as transactions and will be easier to think through. This should also ensure that if there is a starting snapshot set, that the starting snapshot is the furthest that the producer will roll back to.

**Comment by singhpk234** [comment]
Apologies for the delay in updating the pr, i was in the middle of changing my work location.

Have addressed the feedback and moved the changes to SnapshotProducer from BaseTransaction.

@rdblue @jackye1995 please have a look when you get some time.

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 5]
```diff
@@ -615,4 +630,45 @@ private static void updateTotal(
       }
     }
   }
+
+  private Long snapshotIdToRollbackToOnConflict(Long parentSnapshotId) {
+    Long newParentSnapshotId = parentSnapshotId;
+    if (PropertyUtil.propertyAsBoolean(
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:636]: can we move this property check outside this method? So we don't have such a long if block.

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 6]
```diff
@@ -284,6 +284,10 @@ private TableProperties() {}
   public static final String WRITE_AUDIT_PUBLISH_ENABLED = "write.wap.enabled";
   public static final String WRITE_AUDIT_PUBLISH_ENABLED_DEFAULT = "false";
 
+  public static final String ROLLBACK_COMPACTION_ON_CONFLICTS_ENABLED =
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:287]: I think the name feels a bit awkward to me, first it's a write side feature, so I think it's better to make it start with `write.`. Secondly this is to rollback a REPLACE operation, so feels to me it's better to call it something like `write.rollback-rewrite-enabled`?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 7]
```diff
@@ -215,7 +219,18 @@ public Snapshot apply() {
     long sequenceNumber = base.nextSequenceNumber();
     Long parentSnapshotId = parentSnapshot == null ? null : parentSnapshot.snapshotId();
 
-    validate(base, parentSnapshot);
+    try {
+      validate(base, parentSnapshot);
+    } catch (ValidationException ve) {
+      parentSnapshotId = snapshotIdToRollbackToOnConflict(parentSnapshotId);
+      // if no snapshot is found, so that roll backing to it avoids conflicts,
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:226]: this sentence does not make much sense to me... can you rephrase?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 8]
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
...
+        return null;
+      }
+
+      return newParentSnapshotId;
+    } else {
+      return null;
+    }
+  }
+
+  private Boolean shouldRollbackReplaceOnConflict() {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:672]: it already has a default, can be type `boolean`

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 9]
```diff
@@ -73,6 +73,7 @@ enum TransactionType {
   private TableMetadata current;
   private boolean hasLastOpCommitted;
   private final MetricsReporter reporter;
+  private boolean shouldRollbackCompaction;
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:76]: this name should be updated based on the config name

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java:76]: and seems like it's not used?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 10]
```diff
@@ -215,7 +219,18 @@ public Snapshot apply() {
     long sequenceNumber = base.nextSequenceNumber();
     Long parentSnapshotId = parentSnapshot == null ? null : parentSnapshot.snapshotId();
 
-    validate(base, parentSnapshot);
+    try {
+      validate(base, parentSnapshot);
+    } catch (ValidationException ve) {
+      parentSnapshotId = snapshotIdToRollbackToOnConflict(parentSnapshotId);
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:225]: I think the logic that makes more sense in my mind is:

```
try {
  validate(base, parentSnapshot);
} catch (ValidationException ve) {
  if (shouldRollbackReplaceOnConflict()) {
     // do rollback logic 
 } else {
   // rethrow exception
 }
}
```

instead of having the logic to determine the flag as a part of `snapshotIdToRollbackToOnConflict` which is already determining the new parent snapshot.

What do you think?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 11]
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
...
+      return newParentSnapshotId;
+    } else {
+      return null;
+    }
+  }
+
+  private Boolean shouldRollbackReplaceOnConflict() {
+    return PropertyUtil.propertyAsBoolean(
+        base.properties(),
+        TableProperties.COMMIT_ROLLBACK_REPLACE_ON_CONFLICT_ENABLED,
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:675]: I am a bit not sure if this should be a table level property. I would imagine this to enabled only on certain sessions, but the majority of writers still do not rollback. Is there any other option to make this a runtime config instead?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 12]
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
       }
     }
   }
+
+  private Long snapshotIdToRollbackToOnConflict(Long parentSnapshotId) {
+    Long newParentSnapshotId = parentSnapshotId;
+    if (shouldRollbackReplaceOnConflict()) {
+      // add a set snapshot op on top of base to roll back to parent snapshot
+      boolean isCommitSuccessfullyApplied = false;
+      // Update parentSnapshot to it's grandParent
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:639]: you mean ancestor?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 13]
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
...
+        }
+        Snapshot parentSnapshot = base.snapshot(newParentSnapshotId);
+
+        SetSnapshotOperation setSnapshotOp = new SetSnapshotOperation(tempTableOps);
+        setSnapshotOp.rollbackTo(newParentSnapshotId).commit();
+        try {
+          validate(tempTableOps.current(), parentSnapshot);
+          isCommitSuccessfullyApplied = true;
+        } catch (ValidationException validationException) {
+          // swallow the exception for re-trying
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:658]: can we add a log message here, something like "Rollback to %s and commit still failed, ...."

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 14]
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
...
+        // create a tempTableOperation to pass base with rollback to validate of update
+        TableOperations tempTableOps = ops.temp(base);
+        newParentSnapshotId = base.snapshot(newParentSnapshotId).parentId();
+        if (newParentSnapshotId == null) {
+          return null;
+        }
+        Snapshot parentSnapshot = base.snapshot(newParentSnapshotId);
+
+        SetSnapshotOperation setSnapshotOp = new SetSnapshotOperation(tempTableOps);
+        setSnapshotOp.rollbackTo(newParentSnapshotId).commit();
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:653]: this is not within a transaction, wouldn't this actually commit the table?

Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java]: As per my understanding, since this is based on `TableOperations tempTableOps = ops.temp(base);` so it should not commit, we use this for validation to create a new temp metadata which will act as an uncommitted table metadata similar to what's done in transaction.

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java (hunk 12 above)
Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java]: yup

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java (hunk 10 above)
Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java]: sounds good, made the changes

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 15]
```diff
@@ -615,4 +633,47 @@ private static void updateTotal(
       }
     }
   }
+
+  private Long snapshotIdToRollbackToOnConflict(Long parentSnapshotId) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:637]: can we add a javdoc for this to explain the algorithm instead of embedding all the comments inline? Also the method name feels like just getting the ID, but in fact you are already performing the rollback. So maybe an actionable method name like just `rollbackReplaceOnConflictToParent(Long originalParent)`?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 16]
```diff
@@ -199,6 +201,8 @@ protected void validate(TableMetadata currentMetadata, Snapshot snapshot) {}
    */
   protected abstract List<ManifestFile> apply(TableMetadata metadataToUpdate, Snapshot snapshot);
 
+  protected abstract Long startingSnapshotId();
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:204]: I think this has to be non null? For the cases which it is currently null, it's just because the value is missing. But we always know what is the starting snapshot ID of the commit.

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java (hunk 11 above)
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java:676]: Explored a bit more offline, I think maybe it's better to do this through an update in related action APIs, for example:

```
public interface SupportsRollbackReplace<T> {
  
  T rollbackReplaceOnConflict();
}

public interface AppendFiles extends SnapshotUpdate<AppendFiles>, SupportsRollbackReplace<AppendFiles> {
  ...
}
```

So that in core we can do:

```
table.newAppend().rollbackReplaceOnConflict().addFiles(...).commit()
```

And then we can make this a session config or runtime option in engine.

What do you think?

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java:676]: we could potentially make this more generic, for example just add this to `SnapshotUpdate<T>`

```
  T rollback(List<String> operations);
```

Although I don't know what would be the use case of the other kinds of rollback, but seems like a good generalization.

@rdblue do you have any thoughts and further comments regarding this PR?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java (hunk 15 above)
Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java:635]: we are not actually doing a rollback here, but rather trying to see which parent snapshot id we can rollback to, once we figure that out we apply the current on this parent snapshot id we found and commit to the table so the actual rollback is happening in `apply()` hence named this method as it is, please let me know your thoughts, happy to make the changes accordingly

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java (hunk 16 above)
Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java:210]: are you suggesting the `starting snapshot ID` will be the base on which we are applying the snapshot ?
